### ISSUE TITLE
primal return from bwd_diff (or explicit value_and_bwd_diff operator)

### ISSUE THREAD

**Issue by andersjel** [shader-slang/slang#11372]
# Problem Description
 
When needing both f(x) and the gradient ∇f(x), the only current option is to call f and bwd_diff(f) separately, causing the forward pass to run twice. Since bwd_diff already runs the forward pass internally, it would be trivial in principle to also surface the primal return value — analogous to jax.value_and_grad. This would allow a single call to get both, avoiding the redundant pass.

# Preferred Solution
Instead of using `T.Differential` for an _out_ parameter or return type, use `inout DifferentialPair<T>` and write the primal part to the value as part of the forward pass of `bwd_diff(f)`.

# Alternative Solutions
Provide a separate `value_and_bwd_diff` operator that does this.

# Additional context
Jax has https://docs.jax.dev/en/latest/_autosummary/jax.value_and_grad.html to solve this issue.

**Comment by andersjel** [comment]
> **Void-returning functions.** `value_and_bwd_diff(f)` on a void `f` will be a compile error: _"`value_and_bwd_diff` requires a function with a non-void return type; use `bwd_diff` for void-returning functions."_ Erring on the strict side here — relaxing to auto-degenerate later is non-breaking; the reverse isn't.

I think it makes sense to support void returning functions when they have out-parameters (or in-out parameters) at least, in the same way that they are supported by `bwf_diff` and `fwd_diff`. `value_and_bwd_diff(f)` would turn out-parameters into in-out parameters that take the seed as input and writes the primaries during the forward pass.

**Comment by andersjel** [comment]
Hi nv-slang-bot. I am looking forward to trying this out on our project at work. Some comments on the _ValueAndBackwardDerivativeOf_ attribute, and I guess these are for Phase 2: 

- If no `[ValueAndBackwardDerivativeOf(f)]` exists, I would expect the `[BackwardDerivativeOf(f)]` and `f` itself to be called by `value_and_bwd_diff(f)`, instead of falling back to an auto-synthesized derivative. In that way, `[ValueAndBackwardDerivativeOf(f)]` becomes purely a performance optimization.
- As a user, it would be desirable to only specify `[ValueAndBackwardDerivativeOf(f)]`, and then that would be used for `bwd_diff(f)` also, unless there is an explicit `[BackwardDerivativeOf(f)]`.

**Comment by jhelferty-nv** [comment]
@szihs FYI, the bot is pretty noisy here, and may be out over its skis a bit.

@saipraveenb25 I assume you want to be involved here/driving? Assigning you to steer.

**Comment by saipraveenb25** [comment]
Hmm, a couple of things: 
- Since this is a language feature, it should generally go through a language-spec proposal (https://github.com/shader-slang/spec). 
- We recently made a few changes to autodiff, and the JAX style value + grad-fn might already be available through the "apply" method. It's just not officially documented yet since we're still testing. This refactor was completed in https://github.com/shader-slang/slang/pull/9808, but the patch notes are quite messy, so I'll paste just the important bit here:
> ## New Reverse-Mode Auto-Diff Model
> In this system, the reverse-mode derivative of a function is decomposed into three pieces:
> 
> `apply_bwd` — Evaluates the primal function and returns the result along with a minimal context (a small amount of data that is cheap enough to store per call). This is the only piece that runs during the forward pass.
>`remat` (rematerialize) — Given the minimal context and the original function inputs, reconstructs the full context (the complete set of intermediate values needed for back-propagation). This runs at the start of the backward pass.
> `operator()` (back-propagate) — Given the full context and the output differentials, propagates derivatives back to the input differentials.

Essentially, we want to expose something like `apply_bwd(foo)` which will return a tuple of the result and a callable object (i.e. struct with an operator() method) to perform the back-propagation at a later time. In fact, this is how auto-diff currently operates under the hood (when you call `bwd_diff`, it does both the apply and the back-prop one after the other), but the separate pieces are not currently exposed as a user-facing feature since we wanted to go through the language spec process first. 

Will this help solve your issue?

**Comment by andersjel** [comment]
> Will this help solve your issue?

Sure, that sounds usable and gives some freedom beyond what I discussed above with the bots (specifically you can choose not to run a backwards pass after doing the forward pass). 

For the common case of getting both f(x) and ∇f(x), it might get a little ceremonious to use.
